### ISSUE TITLE
experiment with 64-bit ubuntu OS for rpi4

### ISSUE THREAD

**Issue by billimek** [billimek/homelab-infrastructure#10]
See:

* https://ubuntu.com/download/iot/raspberry-pi
* https://wiki.ubuntu.com/EoanErmine/ReleaseNotes
* https://www.reddit.com/r/raspberry_pi/comments/dj8whf/official_ubuntu_release_for_pi_4/

**Comment by billimek** [comment]
There is a pretty bad bug with [USB devices not working](https://bugs.launchpad.net/ubuntu/+bug/1848703) with this release, so will hold-off until this is fixed.

**Comment by onedr0p** [comment]
If I am reading this right, USB issue appears to be fixed in 19.10.1 as per [this issue](https://bugs.launchpad.net/ubuntu/+source/linux-raspi2/+bug/1848790)

Moreover,

https://wiki.ubuntu.com/ARM/RaspberryPi

> Further releases can be found at http://cdimage.ubuntu.com/ubuntu/releases/. Please note that even though these images are labelled "+raspi3" they are compatible with the Raspberry Pi 2 and 3 (and 4 in the case of 19.10).

If wanting 64bit on RPi4 there's also https://github.com/rancher/os but I can feel the need to want Ubuntu instead.

I will experiment over the holiday break.

**Comment by onedr0p** [comment]
There's also using arm64 on HypriotOS or Raspbian with using

`/boot/config.txt`
```
# Enable arm64
arm_64bit=1
```

and then Reboot

```
uname -a
Linux pik3s01 4.19.75-v8+ #1270 SMP PREEMPT Tue Sep 24 18:59:17 BST 2019 aarch64 GNU/Linux
```

**Comment by billimek** [comment]
Successfully installed ubuntu 19.10 for arm64 on `pi4-c`:

```
➜ k get node -o=wide
NAME    STATUS   ROLES    AGE   VERSION         INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION      CONTAINER-RUNTIME
pi4-c   Ready    <none>   68s   v1.16.3-k3s.2   10.2.0.68      <none>        Ubuntu 19.10                     5.3.0-1014-raspi2   containerd://1.3.0-k3s.4
pi4-b   Ready    worker   51d   v1.16.3-k3s.2   10.2.0.59     <none>        Raspbian GNU/Linux 10 (buster)   4.19.75-v7l+        containerd://1.3.0-k3s.4
k8s-4   Ready    worker   68d   v1.16.3-k3s.2   10.2.0.16     <none>        Ubuntu 19.10                     5.3.0-24-generic    containerd://1.3.0-k3s.4
pi4-a   Ready    worker   50d   v1.16.3-k3s.2   10.2.0.17     <none>        Raspbian GNU/Linux 10 (buster)   4.19.75-v7l+        containerd://1.3.0-k3s.4
k3s-2   Ready    worker   68d   v1.16.3-k3s.2   10.2.0.32     <none>        Ubuntu 19.10                     5.3.0-24-generic    containerd://1.3.0-k3s.4
k3s-1   Ready    worker   68d   v1.16.3-k3s.2   10.2.0.31     <none>        Ubuntu 19.10                     5.3.0-24-generic    containerd://1.3.0-k3s.4
k3s-0   Ready    master   68d   v1.16.3-k3s.2   10.2.0.30     <none>        Ubuntu 19.10                     5.3.0-24-generic    containerd://1.3.0-k3s.4
```

The following steps were required:

* in `/etc/netplan/50-cloud-init.yaml`:

```
network:
    ethernets:
        eth0:
            dhcp4: false
            optional: true
    version: 2
    vlans:
        vlan.20:
            id: 20
            link: eth0
            dhcp4: true
            addresses: [10.2.0.1/24]
```

* in `/boot/firmware/nobtcmd.txt` add `cgroup_memory=1 cgroup_enable=memory`:

```
net.ifnames=0 dwc_otg.lpm_enable=0 console=ttyAMA0,115200 console=tty1 root=LABEL=writable rootfstype=ext4 elevator=deadline rootwait fixrtc cgroup_memory=1 cgroup_enable=memory
```

* in `/etc/fstab`, add the following for the USB drive:

```
UUID=7d7baa42-5cd4-4514-a91b-c37040a39fa9 /mnt/usb ext4 defaults 1 2
```

**Comment by billimek** [comment]
After observing for about a week, if all looks good will convert `pi4-a` and `pi4-b` to arm64.

**Comment by onedr0p** [comment]
Were you able to install Ubuntu on the pi completely headless? The instructions I see require a mouse / keyboard to access. 

My guess is we need to mount the usb devices after flashing to copy over the cloud-init changes?

**Comment by billimek** [comment]
Writing some docs for this, will close with that commit!

**Comment by billimek** [comment]
@onedr0p some documentation for provisioning headless arm64 pi4 nodes: https://github.com/billimek/homelab-infrastructure/blob/master/k3s/arm64_pi4_nodes.md

**Comment by onedr0p** [comment]
Looks good, thanks! I am hoping someone out there can make it a bit easier to provision the SD card like [flash](https://github.com/hypriot/flash) does. Tearing down and re-provisioning is not so quick with Ubuntu on RPi it seems :(

**Comment by onedr0p** [comment]
It looks like we're able to use [flash](https://github.com/hypriot/flash)!

[Check it out](https://github.com/onedr0p/k3s-gitops-arm/blob/master/docs/ubuntu.md).

tl;dr:

Pass in a [user-data.yml](https://github.com/onedr0p/k3s-gitops-arm/blob/master/setup/ubuntu/user-data.yml) to the flash command.

```bash
flash \
    --userdata setup/ubuntu/user-data.yml \
    --hostname my-awesome-hostname \
    ~/Downloads/ubuntu-19.10.1-preinstalled-server-arm64+raspi3.img
```

I was also able to get a [ansible playbook](https://github.com/onedr0p/k3s-gitops-arm/tree/master/setup/ansible-ubuntu) for the other tedious things like installing Docker, setting cgroups and getting the RPi [userland](https://github.com/raspberrypi/userland) packages compiled which gives us access to the `/opt/vc/bin/vcgencmd` command.

```bash
devin@pik3s10:~$ uname -a
Linux pik3s10 5.3.0-1014-raspi2 #16-Ubuntu SMP Tue Nov 26 11:18:23 UTC 2019 aarch64 aarch64 aarch64 GNU/Linux

devin@pik3s10:~$ sudo /opt/vc/bin/vcgencmd measure_temp
temp=57.0'C
```

**Comment by billimek** [comment]
This flash approach looks awesome - thank you for sharing it!
